### ISSUE TITLE
Swarm in 1.13 stops listening on published port after turning off healthcheck

### ISSUE THREAD

**Issue by titpetric** [docker/docker#31342]
I'm having a problem where docker swarm would stop listening to published port after health checks would be turned off.

**Steps to reproduce the issue:**

Steps to reproduce from a clean 3 node swarm cluster (all managers):

```
root@swarm1:~# docker node ls
ID                           HOSTNAME  STATUS  AVAILABILITY  MANAGER STATUS
2jrjw0f0t6k2hbf2w0d41db2x *  swarm1    Ready   Active        Reachable
aet822iruebj35gu4vlkn70bb    swarm3    Ready   Active        Leader
l4bdxu989ktteb61cte66dqbc    swarm2    Ready   Active        Reachable
```

I created a `party-swarm` network like this:

```
docker network create \
  --driver overlay \
  --subnet 10.0.0.0/20 \
  --attachable \
  party-swarm
```

I need a simple redis service for the example app:

```
docker service create --replicas 1 --network party-swarm --hostname redis --name redis -p 6379:6379 redis:3.2-alpine
```

And then I create the app like this:

```
docker service create -e PORT=80 \
 --replicas 10 \
 --network party-swarm \
 --name gotwitter \
 -p 80:80 titpetric/gotwitter
```

At this point the service is reachable on port 80 on all nodes.

```
docker service update --no-healthcheck [service]
```

**Describe the results you received:**

At this point, dockerd doesn't listen on port 80 anymore.

**Describe the results you expected:**

I expected dockerd would listen on port 80.

**Additional information you deem important (e.g. issue happens only occasionally):**

I can restore listening on the published port by issuing `docker service update --health-cmd "exit 0" --health-interval 30s gotwitter`

**Output of `docker version`:**

```
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba372
 Built:        Wed Feb  8 06:44:30 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   092cba372
 Built:        Wed Feb  8 06:44:30 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 584
 Running: 260
 Paused: 0
 Stopped: 324
Images: 6
Server Version: 1.13.1
Storage Driver: devicemapper
 Pool Name: docker-8:1-5243831-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 811.1 MB
 Data Space Total: 107.4 GB
 Data Space Available: 106.6 GB
 Metadata Space Used: 22.82 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.125 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.137 (2016-11-30)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: 2jrjw0f0t6k2hbf2w0d41db2x
 Is Manager: true
 ClusterID: 2y5aa4qjf3e6akbclqx2rnx71
 Managers: 3
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 10.55.8.161
 Manager Addresses:
  10.55.8.161:2377
  10.55.8.162:2377
  10.55.8.163:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.6.0-1-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 6
Total Memory: 7.793 GiB
Name: swarm1
ID: 3ER2:7GXI:IZR5:O7M6:HKJT:I22T:LOWM:NVW2:CBGL:5WV3:3B2Y:2YW4
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory limit support
WARNING: No oom kill disable support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

I'm running the swarm on 3 servers, VMs, each with 8GB ram, enough of disk space. I'm trying to scale the service to run 1000 containers and so far haven't been able to go beyond about 750 containers. In order to make it work, I also had to tune some sysctl options, most notably these:

```
sysctl -w net.ipv4.neigh.default.gc_thresh1=8096
sysctl -w net.ipv4.neigh.default.gc_thresh2=8096
sysctl -w net.ipv4.neigh.default.gc_thresh3=8096
```

This was done because dmesg had errors in the form: `neighbour: arp_cache: neighbor table overflow!`. This in turn would cause network connectivity issues to the docker hosts themselves. I suspect this has something to do with the number of docker containers I'm spawning, but the only limits I could discern were that there's some limit about open ports per system (1024), perhaps related with [this gh issue](https://github.com/docker/libnetwork/issues/1522).  That being said, the limits seem to be lower for some reason, and I haven't figured a way to pinpoint the core issue behind them. Even with the health-cmd work-around `docker service ls` is reporting drops in the number of running containers.

```
root@swarm1:~# docker service ls
ID            NAME       MODE        REPLICAS  IMAGE
jgt62g2bsydm  gotwitter  replicated  738/1000  titpetric/gotwitter:latest
syhtbxnz8zhh  redis      replicated  1/1       redis:3.2-alpine
root@swarm1:~# docker service ls
ID            NAME       MODE        REPLICAS  IMAGE
jgt62g2bsydm  gotwitter  replicated  651/1000  titpetric/gotwitter:latest
syhtbxnz8zhh  redis      replicated  1/1       redis:3.2-alpine
```

This may have something to do with the healthcheck timeout option (set by Dockerfile, 10s), but I haven't verified this yet.

The source for the `gotwitter` image is [here](https://github.com/titpetric/books/blob/master/12fa-docker-golang/chapter7/Dockerfile)

**Comment by titpetric** [comment]
I since upgraded the debian strech installs to the latest kernel (from 4.6 to 4.9) and rebooted all swarm nodes. I can verify that the issue persists on the newer version.

```
root@swarm1:~# docker service ls
ID            NAME       MODE        REPLICAS  IMAGE
nwtnkak4yny4  gotwitter  replicated  0/10      titpetric/gotwitter:latest
syhtbxnz8zhh  redis      replicated  1/1       redis:3.2-alpine
```

I wait a few minutes until service becomes 10/10 (dockerd doesn't listen to port 80 at this time).

```
root@swarm1:~# docker service ls
ID            NAME       MODE        REPLICAS  IMAGE
nwtnkak4yny4  gotwitter  replicated  10/10     titpetric/gotwitter:latest
syhtbxnz8zhh  redis      replicated  1/1       redis:3.2-alpine
root@swarm1:~# netstat -pan | grep docker | grep tcp | grep LIST
tcp6       0      0 :::80                   :::*                    LISTEN      497/dockerd
tcp6       0      0 :::2377                 :::*                    LISTEN      497/dockerd
tcp6       0      0 :::7946                 :::*                    LISTEN      497/dockerd
tcp6       0      0 :::6379                 :::*                    LISTEN      497/dockerd
```

After some time (<30s), docker stops listening to port 80:

```
root@swarm1:~# netstat -pan | grep docker | grep tcp | grep LIST
tcp6       0      0 :::80                   :::*                    LISTEN      497/dockerd
tcp6       0      0 :::2377                 :::*                    LISTEN      497/dockerd
tcp6       0      0 :::7946                 :::*                    LISTEN      497/dockerd
tcp6       0      0 :::6379                 :::*                    LISTEN      497/dockerd
root@swarm1:~# docker service ls
ID            NAME       MODE        REPLICAS  IMAGE
nwtnkak4yny4  gotwitter  replicated  10/10     titpetric/gotwitter:latest
syhtbxnz8zhh  redis      replicated  1/1       redis:3.2-alpine
root@swarm1:~# netstat -pan | grep docker | grep tcp | grep LIST
tcp6       0      0 :::80                   :::*                    LISTEN      497/dockerd
tcp6       0      0 :::2377                 :::*                    LISTEN      497/dockerd
tcp6       0      0 :::7946                 :::*                    LISTEN      497/dockerd
tcp6       0      0 :::6379                 :::*                    LISTEN      497/dockerd
root@swarm1:~# netstat -pan | grep docker | grep tcp | grep LIST
tcp6       0      0 :::2377                 :::*                    LISTEN      497/dockerd
tcp6       0      0 :::7946                 :::*                    LISTEN      497/dockerd
tcp6       0      0 :::6379                 :::*                    LISTEN      497/dockerd
root@swarm1:~#
```

Adding the fake health check:

```
root@swarm1:~# docker service update --health-cmd "exit 0" --health-interval 60s --health-timeout 60s gotwitter
```

Docker starts listening on port 80 again.

Updated docker info:

```root@swarm1:~# docker info
Containers: 9
 Running: 4
 Paused: 0
 Stopped: 5
Images: 6
Server Version: 1.13.1
Storage Driver: devicemapper
 Pool Name: docker-8:1-5243831-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 393.9 MB
 Data Space Total: 107.4 GB
 Data Space Available: 107 GB
 Metadata Space Used: 1.499 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.146 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.137 (2016-11-30)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: 2jrjw0f0t6k2hbf2w0d41db2x
 Is Manager: true
 ClusterID: 2y5aa4qjf3e6akbclqx2rnx71
 Managers: 3
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 10.55.8.161
 Manager Addresses:
  10.55.8.161:2377
  10.55.8.162:2377
  10.55.8.163:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.0-1-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 6
Total Memory: 7.792 GiB
Name: swarm1
ID: 3ER2:7GXI:IZR5:O7M6:HKJT:I22T:LOWM:NVW2:CBGL:5WV3:3B2Y:2YW4
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

Also: `docker service scale gotwitter=1000` still floats around 500-800/1000 even with updated healt-cmd/health-timeout=60s. It just doesn't want to go all the way. I suspect I might be limiting myself here with -p, as the containers get added to the `ingress` network, along with my custom `party-swarm` network?

**Comment by aluzzardi** [comment]
Thanks for the super detailed bug report!

Rather than a networking issue, I believe this is a simple matter of "--no-healthcheck" not working at all and thus, with a misbehaving health check that always fail (which seems to be the case here, correct?), swarm will never send traffic from port 80 since no containers appear to be healthy.

I'm basing the above conclusion on the fact that using a --health-cmd that does succeed seems to be fixing the issue

/cc @nishanttotla since you originally implemented the HC feature in Swarm
/cc @dongluochen as you've worked on health check + load balancer in 1.13

**Comment by titpetric** [comment]
Yes, in principle you are right, `--no-healthcheck` has the same result as if you would give `exit 1` into health-cmd, apart from that the command `docker service ls` reports 10/10 (failed health checks would report 0/10 I believe - containers waiting on first health check don't count towards the active service containers).

I'll have the server setup running a while and I'm at your disposal if you would like to run some additional diagnostics or tests. I'm pretty sure that if I would pipe `docker events` into a file, the cause of this might show up there?

**Comment by titpetric** [comment]
FYI, bug is also visible if service is created with the `--no-healthcheck` option to begin with.

**Comment by dongluochen** [comment]
The code didn't cover some variants of healthcheck configuration. This was fixed by https://github.com/docker/docker/pull/30197.

**Comment by thaJeztah** [comment]
@dongluochen https://github.com/docker/docker/pull/30197 is included in 17.03.1, should this issue be closed?

**Comment by titpetric** [comment]
@thaJeztah I still have the swarm up and running, I'll retest it shortly

**Comment by thaJeztah** [comment]
@titpetric packages for `docker-engine` can be found here; http://apt.dockerproject.org/repo/pool/testing/d/docker-engine/

For `docker-ce` (but they're the same, other than a different name); use the regular installation instructions, but use the "test" channel instead of "stable" or "edge"

**Comment by titpetric** [comment]
Thanks, I can't do that however :) I'll have to wait for 17.03.1 to come to the stable tree.

**Comment by dongluochen** [comment]
@thaJeztah I have tested the fix in 17.03.1. You may close this issue.

**Comment by thaJeztah** [comment]
Thanks! Closing, but let us hear if you still have this after upgrading @titpetric
